src/librustc_resolve/lib.rs

```diff
@@ -911,162 +904,156 @@ bitflags! {
-                    type_span: None,
-                });
-            }
-            Some(type_def) => {
-                match type_def.module_def {
-                    None => {
-                        let module = Module::new(parent_link, def_id, kind, external, is_public);
-                        *self.type_def.borrow_mut() = Some(TypeNsDef {
-                            modifiers: modifiers,
-                            module_def: Some(Rc::new(module)),
-                            type_def: type_def.type_def,
-                            type_span: None,
-                        });
-                    }
-                    Some(module_def) => module_def.kind.set(kind),
-                }
-            }
-        }
+        let module = Module::new(parent_link, def, external, is_public);
+        self.type_ns.set(NsDef::create_from_module(Rc::new(module), Some(sp)));
     }
 
     /// Records a type definition.
     fn define_type(&self, def: Def, sp: Span, modifiers: DefModifiers) {
-        debug!("defining type for def {:?} with modifiers {:?}",
-               def,
-               modifiers);
-        // Merges the type with the existing type def or creates a new one.
-        let type_def = self.type_def.borrow().clone();
-        match type_def {
-            None => {
-                *self.type_def.borrow_mut() = Some(TypeNsDef {
-                    module_def: None,
-                    type_def: Some(def),
-                    type_span: Some(sp),
-                    modifiers: modifiers,
-                });
-            }
-            Some(type_def) => {
-                *self.type_def.borrow_mut() = Some(TypeNsDef {
-                    module_def: type_def.module_def,
-                    type_def: Some(def),
-                    type_span: Some(sp),
-                    modifiers: modifiers,
-                });
-            }
-        }
+        debug!("defining type for def {:?} with modifiers {:?}", def, modifiers);
+        self.type_ns.set(NsDef::create_from_def(def, modifiers, Some(sp)));
     }
 
     /// Records a value definition.
     fn define_value(&self, def: Def, sp: Span, modifiers: DefModifiers) {
-        debug!("defining value for def {:?} with modifiers {:?}",
-               def,
-               modifiers);
-        *self.value_def.borrow_mut() = Some(ValueNsDef {
-            def: def,
-            value_span: Some(sp),
-            modifiers: modifiers,
-        });
+        debug!("defining value for def {:?} with modifiers {:?}", def, modifiers);
+        self.value_ns.set(NsDef::create_from_def(def, modifiers, Some(sp)));
     }
 
     /// Returns the module node if applicable.
-    fn get_module_if_available(&self) -> Option<Rc<Module>> {
-        match *self.type_def.borrow() {
-            Some(ref type_def) => type_def.module_def.clone(),
-            None => None,
-        }
-    }
+    fn get_module_if_available(&self) -> Option<Rc<Module>> { self.type_ns.module() }
```

Comment:
Nit: I think we are moving away from these pithy one-liners, in this brave new era of automated formatting (not yet quite upon is). But anyway I think it looks a bit odd to intermingle one-liners like this with the rest of the code here anyway. :)